### PR TITLE
feat: openAPI spec for rudder-server

### PR DESCRIPTION
**Pull request by cisse21**
# Description

	- Added validation for OpenAPI definition in the CI workflow.
	- Introduced a new endpoint `/docs` for serving the OpenAPI specification in HTML format.
	- Generated and embedded OpenAPI specification within the application.

## Linear Ticket

Fixes PIPE-593

## Security

- [ ] The code changed/added as part of this pull request won't create any security issues with how the software is being used.

### CODE REVIEW

File: gateway/handle_lifecycle.go [hunk 1]
```diff
@@ -428,8 +428,14 @@ func (gw *Handle) StartWebHandler(ctx context.Context) error {
 	})
 
 	srvMux.Get("/health", withContentType("application/json; charset=utf-8", app.LivenessHandler(gw.jobsDB)))
-
 	srvMux.Get("/", withContentType("application/json; charset=utf-8", app.LivenessHandler(gw.jobsDB)))
+	srvMux.Get("/api-spec",
```
Comment (fracasula) [inline gateway/handle_lifecycle.go:432]: what about `/docs` or `/openapi`?

**Comment by cisse21** [comment]
@coderabbitai review

File: gateway/handle_lifecycle.go [hunk 2]
```diff
@@ -428,8 +428,14 @@ func (gw *Handle) StartWebHandler(ctx context.Context) error {
...
+		func(w http.ResponseWriter, r *http.Request) {
+			w.Header().Set("Content-Type", "text/html; charset=utf-8")
+			_, _ = w.Write(openApiSpec)
+		},
+	)
```
[the remark this answers is not in the document]
Comment (cisse21) [reply gateway/handle_lifecycle.go:437]: Where was the root endpoint removed? I see no deletions in the file

File: gateway/gateway_integration_test.go [hunk 3]
```diff
@@ -83,6 +83,56 @@ func TestWebhook(t *testing.T) {
...
+		err := runGateway(ctx, gwPort, postgresContainer, bcServer.URL, transformerContainer.TransformURL, t.TempDir())
+		if err != nil {
+			t.Logf("rudder-server exited with error: %v", err)
+		}
+		return err
+	})
+
+	url := fmt.Sprintf("http://localhost:%d", gwPort)
+	health.WaitUntilReady(ctx, t, url+"/health", 60*time.Second, 10*time.Millisecond, t.Name())
+	// send an event
```
Comment (achettyiitr) [inline gateway/gateway_integration_test.go:126]:
```suggestion
```

File: gateway/gateway_integration_test.go [hunk 4]
```diff
@@ -83,6 +83,56 @@ func TestWebhook(t *testing.T) {
...
+	req, err := http.NewRequest(http.MethodGet, url+"/docs", nil)
+	require.NoError(t, err)
+	resp, err := (&http.Client{}).Do(req)
```
Comment (achettyiitr) [inline gateway/gateway_integration_test.go:129]: [optional]: 
```suggestion
	resp, err := http.Get(url + "/docs")
```

**Review by achettyiitr — APPROVED** [review]

File: Makefile [hunk 5]
```diff
@@ -3,6 +3,7 @@
 GO=go
 LDFLAGS?=-s -w
 TESTFILE=_testok
+MOUNT_PATH= /local
```
Comment (fracasula) [inline Makefile:6]: [optional] not sure if it gets trimmed already in some other stages, probably when mounting.

```suggestion
MOUNT_PATH=/local
```

File: gateway/gateway_integration_test.go [hunk 6]
```diff
@@ -83,6 +83,53 @@ func TestWebhook(t *testing.T) {
...
+func TestDocsEndpoint(t *testing.T) {
+	bcServer := backendconfigtest.NewBuilder().
+		WithWorkspaceConfig(
+			backendconfigtest.NewConfigBuilder().
+				WithSource(
+					backendconfigtest.NewSourceBuilder().
+						WithID("source-1").
+						WithWriteKey("writekey-1").
+						WithSourceCategory("webhook").
+						WithSourceType("SeGment").
```
Comment (fracasula) [inline gateway/gateway_integration_test.go:95]:
```suggestion
						WithSourceType("my_source_type").
```

File: gateway/openapi.yaml [hunk 7]
```diff
@@ -0,0 +1,908 @@
+openapi: 3.0.1
```
Comment (fracasula) [inline gateway/openapi.yaml:1]: [optional] we could use the latest as well i.e. 3.0.3

File: gateway/gateway_integration_test.go [hunk 8]
```diff
@@ -83,6 +83,53 @@ func TestWebhook(t *testing.T) {
...
+		return err
+	})
+
+	url := fmt.Sprintf("http://localhost:%d", gwPort)
+	health.WaitUntilReady(ctx, t, url+"/health", 60*time.Second, 10*time.Millisecond, t.Name())
+	resp, err := http.Get(url + "/docs")
+	require.NoError(t, err, "it should be able to get the docs")
+	require.Equal(t, http.StatusOK, resp.StatusCode)
+	func() { kithttputil.CloseResponse(resp) }()
+	require.Equal(t, resp.Header.Get("Content-Type"), "text/html; charset=utf-8")
```
Comment (fracasula) [inline gateway/gateway_integration_test.go:130]: We could have a `require.Contains` on the response body as well. Just something to make sure it has some content that it makes sense.

**Review by fracasula — APPROVED** [review]
Approved with suggestions :+1:

**Pull request merged**